### DISCUSSION TITLE
更新日志 - 22.03.6 - 20240119

### DISCUSSION THREAD

**Discussion by jjm2473** [istoreos/istoreos#1281] [Announcements]
__本地化网关__

# 内核
1. 修复 r8169 驱动死锁问题（x86 已提前发过一版） #1257

# 系统
1. DHCP/DNS 增加“标签”功能，用于指派不同的网关 和/或 DNS 给不同的内网主机。此功能对于有旁路由的情况应该很有用，不管iStoreOS作为主路由还是旁路由，只要iStoreOS提供DHCP服务，都能利用此功能区分不同主机的上网需求。例如可以默认指派主路由作为网关，但对于特定内网主机则指派旁路由作为其网关。后台具体位置“网络”-“DHCP/DNS”-“标签”，配置前仔细阅读其中的说明。简单的使用教程请看 issue  #1237
2. 优化磁盘挂载功能，增加“关键挂载点”选项（在每个挂载点的高级设置中），用于系统启动过程中等待特定磁盘分区准备好，避免程序启动时硬盘还没挂载的情况；默认最多等待5秒，可在“挂载点”的“全局设置”中自定义。此外，增加 Hook 机制，允许软件包自动添加“关键挂载点”，而无需用户配置。目前 Docker 已使用此 Hook，数据目录及其祖先目录将自动视为“关键挂载点”。（此功能可替代杂项设置中的“启动”配置）
3. 修复某些情况下固件升级失败的问题。如果之前遇到升级失败，参考 issue 解决 #1200

# 应用
* QuickStart: 修复自定义 CheckBox 组件不被屏幕阅读器识别的问题 #1272

# 设备
* x86_64：手动升级固件时，校验用户上传的固件与运行中版本是否匹配EFI模式，不匹配将警告，防止用户刷坏引导程序
* rk35xx：对于部分机型（例如T68m）的 USB OTG 不支持自动切换，开机默认 Host 模式 #1255

**Comment by jjm2473** [comment]
关于passwall的问题：
1. 从这里下载passwall，以避免openssl版本不匹配 https://github.com/bcseputetto/Are-u-ok/releases
2. passwall默认开机60秒以后才开始运行，这是非常不合理的，尤其对于x86设备，到passwall的高级设置里面，把“开机时延时启动”设置成5以内的数字就行了
3. 如果不确定节点是否支持ipv6，那么，先在后台的“网络”-“DHCP/DNS”-“高级设置”，勾选“过滤 IPv6 AAAA 记录”，以避免ipv6转发。
4. 为避免DNS缓存污染，或者DNS对应的ip没加入ipset，客户端（手机电脑等）如果访问外网失败就重新联网试试。最好软路由重启以后，把wifi路由器也重启下，强制所有设备重新联网。
5. 负载均衡启动失败，重装haproxy试试：软件包那边点一下更新列表，然后搜索haproxy，卸载重装。可能之后要重装下passwall。
6. 如果进行以上操作以后还有问题，再来反馈

**Reply by aiqingvsyy** [reply]
你好  请问虚拟机实例不跟软路由启动而启动是什么情况  是我设置的不对还是一个bug

**Reply by jjm2473** [reply]
这个要问插件作者了

**Comment by jjm2473** [comment]
## 本地化网关的简单教程

首先了解下 “DHCP 选项” 的语法：
* “3,192.168.100.2”表示指派 192.168.100.2 作为网关。（注意3和英文逗号）
* “6,192.168.100.2”表示指派 192.168.100.2 作为 DNS 服务器。（注意6和英文逗号）
* 其他选项不常用就不解释了

典型使用场景（必须iStoreOS提供DHCP服务，以下皆假设主路由IP是192.168.100.1，旁路由IP是192.168.100.2，掩码都是24位）：
* iStoreOS作为旁路由，连接了办公网VPN，内网部分主机需要访问VPN网络（假设这部分主机叫工作机），大部分主机都不需要
  1. 打开“网络”-“接口”，在“lan”后面点击编辑，在弹出的对话框选择“DHCP 服务器”-“高级设置”，其中的“DHCP 选项”增加两条“3,192.168.100.1”和“6,192.168.100.1”
  2. 打开“网络”-“DHCP/DNS”-“标签”，在“添加”按钮左边输入“t_work”，点击“添加”按钮，然后也是“DHCP 选项”输入两条“3,192.168.100.2”和“6,192.168.100.2”，保存应用
  3. 打开“网络”-“DHCP/DNS”-“静态地址分配”，点击“添加”（如果之前已经配置了工作机，那编辑相应条目即可），输入主机名称（可以用英文大小写，数字，下划线，必填），选择工作机的MAC；如果不需要固定IP，“IPv4 地址”可以选择未指定；最关键的是在“标签”选项中选择“t_work”（如果没出现，保存下刷新页面再回来选）。然后保存应用即可。对于每台工作机都要添加。内网主机重新插网线或连wifi，应该就会获取到指定的网关和DNS了。
* iStoreOS作为主路由，未连接办公网VPN，旁路由连接了办公网VPN，内网部分主机需要访问VPN网络（假设这部分主机叫工作机），大部分主机都不需要。步骤跟上面完全一样，参数都不用改，但是跳过第1步（因为默认网关就是主路由）。

**Reply by jjm2473** [reply]
教程转发到 https://github.com/istoreos/istoreos/discussions/1944 ，方便查找

**Comment by bcseputetto** [comment]
群里部分用户出现了更新后接口丢失防火墙区域导致断网的问题，在这里提醒！
![image](https://github.com/istoreos/istoreos/assets/29139573/e50000bf-5490-4b24-adc0-46c58015ecc3)
如果也出现类似问题，请到路由器后台，将WAN接口的防火墙区域重新设置到WAN区域内。
应该就可以解决了，不需要重刷固件。如果有其它问题也可以反馈

**Reply by jjm2473** [reply]
奇怪，最近没改uci配置啊

**Reply by jjm2473** [reply]
我测试了升级或者重置都没出现防火墙区域丢失的问题。问一下安装了什么插件

**Reply by bcseputetto** [reply]
我没有详细测试，不过从这位用户有VAP开头的接口来看，装有访问AP/光猫，不知是否有关

**Reply by jjm2473** [reply]
访问AP/光猫只是会操作vap相关的那些接口，而且每次开机都会执行一样的操作，不会因为升级而有什么不同

**Comment by bcseputetto** [comment]
关于iStoreOS旧版本升级到新版本后passwall的haproxy(负载均衡)无法使用的问题。经过我的测试发现是某个软件包在升级后不见了导致的。先从解决办法说起。从软件包重新安装 `libpcre` 即可解决
也可以从终端里执行下面的命令安装解决
```console
opkg update
opkg install libpcre
```


---
以下是测试和复现过程
```console
 -------------------------------------------------------
 iStoreOS 22.03.5, 2023122916
 -------------------------------------------------------
root@iStoreOS:~# haproxy
HAProxy version 2.4.22-1 2023/11/26 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2026.
Known bugs: http://www.haproxy.org/bugs/bugs-2.4.22.html
Running on: Linux 5.10.176 #0 SMP Fri Dec 29 07:18:08 2023 x86_64
Usage : haproxy [-f <cfgfile|cfgdir>]* [ -vdVD ] [ -n <maxconn> ] [ -N <maxpconn> ]
        [ -p <pidfile> ] [ -m <max megs> ] [ -C <dir> ] [-- <cfgfile>*]
        -v displays version ; -vv shows known build options.
        -d enters debug mode ; -db only disables background mode.
        -dM[<byte>] poisons memory with <byte> (defaults to 0x50)
        -V enters verbose mode (disables quiet mode)
        -D goes daemon ; -C changes to <dir> before loading files.
        -W master-worker mode.
        -q quiet mode : don't display messages
        -c check mode : only check config files and exit
        -n sets the maximum total # of connections (uses ulimit -n)
        -m limits the usable amount of memory (in MB)
        -N sets the default, per-proxy maximum # of connections (0)
        -L set local peer name (default to hostname)
        -p writes pids of all children to this file
        -de disables epoll() usage even when available
        -dp disables poll() usage even when available
        -dS disables splice usage (broken on old kernels)
        -dG disables getaddrinfo() usage
        -dR disables SO_REUSEPORT usage
        -dr ignores server address resolution failures
        -dV disables SSL verify on servers side
        -dW fails if any warning is emitted
        -dD diagnostic mode : warn about suspicious configuration statements
        -sf/-st [pid ]* finishes/terminates old pids.
        -x <unix_socket> get listening sockets from a unix socket
        -S <bind>[,<bind options>...] new master CLI

root@iStoreOS:~# opkg list-installed |grep libpcre
libpcre - 8.45-1
libpcre2 - 10.37-1
```
以上是升级前的`haproxy`和`libpcre`相关软件包的情况

```console
 -------------------------------------------------------
 iStoreOS 22.03.6, 2024011910
 -------------------------------------------------------
root@iStoreOS:~# haproxy
Error loading shared library libpcreposix.so.0: No such file or directory (needed by /usr/sbin/haproxy)
Error loading shared library libpcre.so.1: No such file or directory (needed by /usr/sbin/haproxy)
Error relocating /usr/sbin/haproxy: pcre_version: symbol not found
Error relocating /usr/sbin/haproxy: pcre_exec: symbol not found
Error relocating /usr/sbin/haproxy: pcre_config: symbol not found
Error relocating /usr/sbin/haproxy: pcre_compile: symbol not found
Error relocating /usr/sbin/haproxy: pcre_study: symbol not found
Error relocating /usr/sbin/haproxy: pcre_free_study: symbol not found
Error relocating /usr/sbin/haproxy: pcre_free: symbol not found
root@iStoreOS:~# opkg list-installed |grep libpcre
libpcre2 - 10.37-1
root@iStoreOS:~# opkg update
Downloading https://istore.linkease.com/repo/all/compat/Packages.gz
Updated list of available packages in /var/opkg-lists/istore_compat
Downloading https://istore.linkease.com/repo/all/compat/Packages.sig
Signature check passed.
Downloading https://mirrors.tuna.tsinghua.edu.cn/openwrt/releases/22.03.6/targets/x86/64/packages/Packages.gz
Updated list of available packages in /var/opkg-lists/openwrt_core
Downloading https://mirrors.tuna.tsinghua.edu.cn/openwrt/releases/22.03.6/targets/x86/64/packages/Packages.sig
Signature check passed.
Downloading https://mirrors.tuna.tsinghua.edu.cn/openwrt/releases/22.03.6/packages/x86_64/base/Packages.gz
Updated list of available packages in /var/opkg-lists/openwrt_base
Downloading https://mirrors.tuna.tsinghua.edu.cn/openwrt/releases/22.03.6/packages/x86_64/base/Packages.sig
Signature check passed.
Downloading https://mirrors.tuna.tsinghua.edu.cn/openwrt/releases/22.03.6/packages/x86_64/luci/Packages.gz
Updated list of available packages in /var/opkg-lists/openwrt_luci
Downloading https://mirrors.tuna.tsinghua.edu.cn/openwrt/releases/22.03.6/packages/x86_64/luci/Packages.sig
Signature check passed.
Downloading https://mirrors.tuna.tsinghua.edu.cn/openwrt/releases/22.03.6/packages/x86_64/packages/Packages.gz
Updated list of available packages in /var/opkg-lists/openwrt_packages
Downloading https://mirrors.tuna.tsinghua.edu.cn/openwrt/releases/22.03.6/packages/x86_64/packages/Packages.sig
Signature check passed.
Downloading https://mirrors.tuna.tsinghua.edu.cn/openwrt/releases/22.03.6/packages/x86_64/routing/Packages.gz
Updated list of available packages in /var/opkg-lists/openwrt_routing
Downloading https://mirrors.tuna.tsinghua.edu.cn/openwrt/releases/22.03.6/packages/x86_64/routing/Packages.sig
Signature check passed.
Downloading https://mirrors.tuna.tsinghua.edu.cn/openwrt/releases/22.03.6/packages/x86_64/telephony/Packages.gz
Updated list of available packages in /var/opkg-lists/openwrt_telephony
Downloading https://mirrors.tuna.tsinghua.edu.cn/openwrt/releases/22.03.6/packages/x86_64/telephony/Packages.sig
Signature check passed.
root@iStoreOS:~# opkg install libpcre
Installing libpcre (8.45-3) to root...
Downloading https://mirrors.tuna.tsinghua.edu.cn/openwrt/releases/22.03.6/packages/x86_64/base/libpcre_8.45-3_x86_64.ipk
Configuring libpcre.
root@iStoreOS:~# haproxy
HAProxy version 2.4.22-1 2023/11/26 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2026.
Known bugs: http://www.haproxy.org/bugs/bugs-2.4.22.html
Running on: Linux 5.10.201 #0 SMP Tue Jan 16 11:33:34 2024 x86_64
Usage : haproxy [-f <cfgfile|cfgdir>]* [ -vdVD ] [ -n <maxconn> ] [ -N <maxpconn> ]
        [ -p <pidfile> ] [ -m <max megs> ] [ -C <dir> ] [-- <cfgfile>*]
        -v displays version ; -vv shows known build options.
        -d enters debug mode ; -db only disables background mode.
        -dM[<byte>] poisons memory with <byte> (defaults to 0x50)
        -V enters verbose mode (disables quiet mode)
        -D goes daemon ; -C changes to <dir> before loading files.
        -W master-worker mode.
        -q quiet mode : don't display messages
        -c check mode : only check config files and exit
        -n sets the maximum total # of connections (uses ulimit -n)
        -m limits the usable amount of memory (in MB)
        -N sets the default, per-proxy maximum # of connections (0)
        -L set local peer name (default to hostname)
        -p writes pids of all children to this file
        -de disables epoll() usage even when available
        -dp disables poll() usage even when available
        -dS disables splice usage (broken on old kernels)
        -dG disables getaddrinfo() usage
        -dR disables SO_REUSEPORT usage
        -dr ignores server address resolution failures
        -dV disables SSL verify on servers side
        -dW fails if any warning is emitted
        -dD diagnostic mode : warn about suspicious configuration statements
        -sf/-st [pid ]* finishes/terminates old pids.
        -x <unix_socket> get listening sockets from a unix socket
        -S <bind>[,<bind options>...] new master CLI
```
以上是升级后的`haproxy`和`libpcre`相关软件包情况。
以及重新安装上`libpcre`后就可以正常运行了

**Reply by jjm2473** [reply]
好像是因为wget的依赖从libpcre变成了libpcre2，所以libpcre就没集成了

**Reply by jjm2473** [reply]
haproxy重新安装的话，应该会自动安装 libpcre

**Reply by jjm2473** [reply]
之后固件会加回libpcre，避免之前安装的软件包的依赖不满足

**Comment by pikelu** [comment]
更新到iStoreOS 22.03.6 2024011910固件后，通过ipv6地址远程桌面连接不上，防火墙设置转发已选择了接受，请楼主修复这个bug，感谢！

**Reply by bcseputetto** [reply]
我测试了一下没有问题，而且你对防火墙的用法是错误的，不能整个接受。有需要开放的IPv6端口请使用**防火墙通信规则**。
![image](https://github.com/istoreos/istoreos/assets/29139573/a54a6bd3-b408-4a94-bf5e-190ae2136871)
防火墙区域WAN的入站和转发**保持拒绝**
接下来是OpenWrt作为主路由时，IPv6防火墙的正确设置方法。
来到防火墙，通信规则，拉到最底部左下角点击【添加】，在弹出的新窗口可以在高级设置里限制地址族为【仅IPv6】
![image](https://github.com/istoreos/istoreos/assets/29139573/ee2fe3a5-1ef9-4ea9-b232-9824008adb3e)
然后我们回到常规设置，接下来分两种情况，一种是开放路由器本机IPv6的端口，详见下图
![IPv6_Router](https://github.com/istoreos/istoreos/assets/29139573/09403991-21ef-462f-bb80-404e5a1fcbd2)
接着是路由器LAN后面设备开放IPv6端口的做法，详见下图
![IPv6_LAN](https://github.com/istoreos/istoreos/assets/29139573/732f414c-c91d-4b73-a1e7-e3339b7d8fee)
可以看到和开放路由器本机端口不一样的地方有**目标区域的选择**，还有**需要填写目标地址**，以及**动态前缀场景下的目标地址写法**。
然后让我来**实际开放Windows电脑的3389端口，实际测试一下连通性**。
![image](https://github.com/istoreos/istoreos/assets/29139573/a9a851f2-c36b-429b-bc05-ceb461f0be0f)
首先是防火墙通信规则，开放我Windows电脑**后缀的一部分为【8945:3373】的IPv6的3389**
接下来使用测试网站检查该IPv6的**3389连通性**
![image](https://github.com/istoreos/istoreos/assets/29139573/d40d4d99-0225-4152-ba78-b47ca37a636d)
注意看**该后缀的一部分为【8945:3373】的IPv6地址的3389连通性是百分之百**。
所以最新的iStoreOS的防火墙是**没有问题的**，**没有BUG**。
顺便贴一下 我用来测试IPv6连通性的网站
https://www.itdog.cn/tcping_ipv6/
注意，只能测试TCP，UDP不能测

**Reply by pikelu** [reply]
感谢大神的指导！

**Reply by bcseputetto** [reply]
对于Windows的SLAAC后缀固定问题做一下补充，由于Windows隐私问题，默认启用了RandomizeIdentifiers
详见 https://songxwn.com/ipv6-Windows/
如果想要关闭该功能，以固定IPv6后缀，在Windows Powershell里执行
```Powershell
Set-NetIPv6Protocol -RandomizeIdentifiers Disabled
```

**Reply by jjm2473** [reply]
如果是要外网直接访问内网设备的公网ipv6，那应该配置通信规则允许转发。或者配置端口转发后用路由器的公网ipv6来访问

**Comment by yiduobianhua** [comment]
为啥我的 R4S 刷入istore 系统直接不启动啊  难道是因为是1G版本的吗

**Reply by jjm2473** [reply]
目前没设备测试，但是根据网友的反馈，可以先刷 21.02.3 最后一个版本，然后再在线升级到最新版本就行了。

**Reply by yiduobianhua** [reply]
好的 我晚上清下tf卡分区  刷上去试试看  如果可以我提供一下 ttl日志

**Reply by jjm2473** [reply]
嗯，有ttl就更好了，开个新issue或者在相关issue把ttl的日志贴一下

**Reply by yiduobianhua** [reply]
我的R4S 坏了  [大哭][大哭][大哭]    已经送去修了  能修好的话  回来立马贴上日志

**Comment by yuanxiang2015** [comment]
更新今天的版本之后，KVM虚拟机创建的win10无法启动，提示如下：
libvirt 错误- Cannot get interface MTU on 'br-vmf': No such device
上一个版本是正确的，设备是x86

**Reply by jjm2473** [reply]
'br-vmf'很明显应该就是kvm自己创建的，看看前面还有没有其他错误，也检查下系统日志和内核日志

**Comment by lhw174780322** [comment]
软路由作为旁路由使用，生级以后发现科学上网都不好使了，用任何翻墙软件链接节点，在软理由⬆️都可以ping通google，但是其他链接软路由的下级设备都不好使。在电脑上显示网址不安全。哪位大神可以指点下

**Reply by OcrSu** [reply]
我是旁路由用来做adguardhome + smartdns，更新之后无法上网，经过排查后旁路由自己向adguardhome + smartdns 发送dns请求是没问题的。主路由和其他设备使用旁路由地址发送DNS请求都是不行的。
目前怀疑是 DHCP/DNS 增加“标签”功能 这个导致的，需要重新配置一下才能行。现在已经主路由换回运营商DNS了.
有没有大佬指导一下QWQ，过几天我再自己研究下怎么搞，

**Reply by jjm2473** [reply]
“标签”只是个前端ui而已，后端完全没变化，dnsmasq本来就支持的。如果你都没配置“标签”，那不可能有影响的。

**Reply by OcrSu** [reply]
那我就不知道了，我更新以后就出bug了。QWQ我啥也没动

**Reply by OcrSu** [reply]
假如我的主路由后台为192.168.2.1  |  iStoreOS旁路由后台为192.168.2.100
在旁路由中安装 adguardhome + smartdns
ADH配置为 使用53端口替代dnsmasq， 上游DNS服务器填 127.0.0.1:8054
SmartDNS配置为： 端口设置为8054，关闭自动设置Dnsmasq选项， 添加常见DNS服务器地址‘

最后在主路由后台设置DNS地址为192.168.2.100。
此时我的电脑网卡信息显示：网关为192.168.2.1， DNS服务器地址为192.168.2.100

理论上来说应该是一切正常，我更新前也是这样的配置。
但更新后 主路由和我的电脑手机就无法正常使用DNS解析这个功能了，ADH后台看不到任何DNS请求。

但如果将旁路由的DNS地址设置为127.0.0.1，旁路由自己可以正常使用ADH+SmartDNS的DNS解析服务，在ADH后台日志可以看到有使用记录。

#请问一下大佬，能不能帮忙看一下我的配置哪里出问题了QWQ。
